### ISSUE TITLE
Redefine `CARGO_TARGET_DIR` to be only an artifacts directory

### ISSUE THREAD

**Issue by kornelski** [rust-lang/cargo#14125]
### Problem

There are a couple of issues with the `CARGO_TARGET_DIR` that are seemingly in conflict with each other:

1. Users don't like that the `target` dir is huge, and that there are many copies of `target` directories scattered around disk for individual projects. Multiple locations complicate excluding them from backups and full-disk search, cleanup of the temp files, moving temp files to dedicated partitions, out of slow network drives or container mounts, etc. Users would prefer a central location to ease management of the temp files, and also to dedupe/reuse dependencies across many projects.

2. People (and tools) are relying on a relative `./target` directory being present to copy or run built files out of there. Additionally, users may not want to configure a shared `CARGO_TARGET_DIR` due to risk of file name conflicts between projects.

However, the dilemma between 1 and 2 exists only because Cargo uses `CARGO_TARGET_DIR` for two different roles:

1. A cache for all intermediate build products (a place where crates.io crates are built, where compiler-private temp files are) which aren't project-specific, and/or files that users don't need to access directly.
2. A location for user-facing final build products (artifacts) that users expect to be there and need to access.

### Proposed Solution

So to satisfy both uses, I suggest to change the thinking about what the role of `CARGO_TARGET_DIR` should be. Instead of thinking *where* to put the same huge all-purpose mixed `CARGO_TARGET_DIR`, think how to deduplicate and slim `CARGO_TARGET_DIR`, and move everything non-user-facing out of it. 

Instead of merging or sharding the `CARGO_TARGET_DIR` as-is with all of its current content, and adding `--artifact-dir` as a separate place where final products are being copied to — make `CARGO_TARGET_DIR` to be the artifact dir (without copying).

As long as the `CARGO_TARGET_DIR` dir is the place for *all* of the build files, of all crates including all the crates.io and local builds, with all the caches, all the temp junk, then this is going to be a problematic large directory that needs to be managed. But if the purpose of the `./target` dir was changed to be only for user-facing files (files that users can name, and would access via `./target` path themselves), then this directory would be relatively small, with a good reason to stay workspace-relative.

### What isn't an intermediate build product? (and should stay in `./target`)

* linked (and stripped) binaries of the current workspace, including binaries for the examples,
* libraries of the current workspace as `.a`/`.so`, where `lib.crate-type` calls for them. Possibly `.rlib`/`.rmeta` in the future if there's a stable ABI.
* linked binaries for tests and benches of the current workspace (to make it easy to launch them under a debugger/profiler, and so they can use relative file paths to read workspace assets).
* debug symbols for all of the above.
* `.d` files for all of the above (so that IDEs and other build systems know when to rebuild the artifacts).
* if Cargo adds some "staging" directory (a non-private `OUT_DIR` for `build.rs`), then for build scripts belonging to the current workspace it would be inside `./target` as well.

So generally files that users build intentionally, and may want to access directly (run themselves, or package up for distribution) and files that users may need configure their IDE and debugger to find inside the project.

Crates in `[patch.crates-io]` with a `path` are a gray area, an might also have their artifacts included in the `./target` dir (but in some way that avoids clobbering workspaces' files).

### What isn't a final build product, and doesn't belong to `./target`:

* anything related to building crates from crates.io, or any other registry (packages with `source = "registry+…"`)
* all `.fingerprint` and `incremental` dir content of all crates. These are implementation details of the compiler, and nobody should be accessing these directly via `./target/…`.
* `.o` files. Users are not supposed to use them directly either (Rust has static libs for this).
* proc macro libs. They're not useful without rustc present.

All of these should be built in some other shared build cache dir (one that is **not** inside `CARGO_TARGET_DIR`), configurable by a new option/env var. 

Registry dependencies would get unique paths derived from rustc version + package IDs + enabled features (so that different crates using different features don't invalidate each others' caches all the time). This would enable sharing built crates.io dependencies across all projects for the same local user, without also causing local workspaces to clobber each others' `CARGO_TARGET_DIR/profile/product` paths. Temp directories for local projects would need some hashed paths in the shared build/temp dir too.

## Advantages

* Such split removes about 90% of the weight from `./target` dirs (for `cargo` itself, it makes `./target/debug` with binaries and tests take 415MB, instead of 4.2GB). This makes cleanup of all the scattered `target` dirs less of a pressing problem.
* The `./target` keeps relatively few files, and removes high-frequency-churning files out of it, which makes it less of a problem for real-time disk indexing (like search and backups on macOS). 
* I/O latency of `./target` stops being critical for build speeds, unlike I/O of the incremental cache and rewrites of thousands of `.o` files. It becomes feasible to have project directory on a network drive without overriding `CARGO_TARGET_DIR` (network filesystems are used by non-Linux systems where tools like Vagrant and Docker have to run full-fat VMs, and can't cheaply share the file system).
* It makes `./target` contain only workspace-unique files, which makes it justified for every workspace to have one. 
* It enables moving registry deps to a shared build directory, without side effect of local projects overwriting each others' files. Sharing of dependencies matches users' expectation that the same dependencies shouldn't be redundantly rebuilt for each local project.
* It's almost entirely backwards compatible. Users can get the benefits without breaking their existing workflows, post-build scripts, and integrations. It doesn't invalidate documentation/books/tutorials that refer to `target/release/exe` etc.
* It could be the default behavior, so it could benefit all users without friction of adding `--artifact-dir` or `.cargo/config`.

**Comment by kornelski** [comment]
In terms of incompatibilities, only a few come to my mind: 

* Hacks that recursively search `./target` looking for a file built by some build script in its `OUT_DIR`. Ideally this should be solved by adding proper support for building arbitrary assets in Cargo, but for the sake of keeping compatibility with `CARGO_TARGET_DIR`, this could be made to work by keeping `OUT_DIR`s as subdirectories or symlinks in `./target`.

* Binaries trying to link to `.so`/`.dylib` of their Rust dependencies, where the deps are built by Cargo, and are not installed on the system. This is currently basically unsupported in Cargo, because there's no way to set `rpath` to anything that works (absolute path is too build-specific, and relative doesn't work for tests and examples that are in a different dir than bins). But if users are hacking around it with custom post-build steps, they may expect to find the 3rd party shared libs in `./target`. This again would be better solved by having Rust/Cargo aware of dylib dependencies, and copy/hardlink dylibs as needed and set rpath accordingly. Not many deps have `crate-type = "cdylib"`, so these also could be kept/symlinked in `./target` for back-compat.

* Build containers could have very strict read-only filesystems with only `CARGO_TARGET_DIR` and `CARGO_HOME` subdirs being writeable. This could be made backwards-compatible by reverting back to the current everything-in-one-place behavior when `CARGO_TARGET_DIR` is set, and picking other env vars for configuring locations of the trimmed `./target` and the intermediate products cache dir.

**Comment by epage** [comment]
@poliorcetics from https://github.com/rust-lang/rfcs/issues/3664#issuecomment-2183534143

> Yes, the issue should be moved to cargo I think.
> 
> I'm not convinced at all this won't break backwards compatibility in some way.
> 
> > It makes ./target contain only workspace-unique files, which makes it justified for every workspace to have one.
> 
> And I don't want one in _any_ cargo project while still keeping isolation, which is entirely different from what you are proposing.
> 
> > It enables moving registry deps to a shared build directory, without side effect of local projects overwriting each others' files. Sharing of dependencies matches users' expectation that the same dependencies shouldn't be redundantly rebuilt for each local project.
> 
> Once again, the RFC I wrote and original issue I inspired myself from _do not_ ask for that, it asks for the opposite: myself and many others want _separate_ targets dirs for every project.
> 
> There are probably as many reasons as there are users for it but common ones are different sets of features amongst projects, sharing of build caches for specific projects, CI builds wanting to separate projects for security, or one project pinning 1.2.3 in a dep B of a dep A and the other project pinning 1.2.4: A can have the same version for both but it's dependencies won't and cargo is not made to handle the case at the moment.
> 
> You are fundamentally solving a different issue, one that the RFC I posted is _not_ trying to solve.

**Comment by epage** [comment]
Overall, I see this as a solution alternative to #6790 and had recommended we have that conversation there (or on internals).

> What isn't an intermediate build product? (and should stay in ./target)

This is likely going to be the most difficult topic to work through and we'll need to make sure we get wide input on this from #6790 users and others. 

> Registry dependencies would get unique paths derived from rustc version + package IDs + enabled features (so that different crates using different features don't invalidate each others' caches all the time). This would enable sharing built crates.io dependencies across all projects for the same local user, without also causing local workspaces to clobber each others' CARGO_TARGET_DIR/profile/product paths. Temp directories for local projects would need some hashed paths in the shared build/temp dir too.

imo this is out of scope for this proposal (see #5931) and we should keep this focused so as not to get distracted.

**Comment by epage** [comment]
I see this as a re-framing of the problem, addressing #6790 and rust-lang/rfcs#3371

Instead of us defining a new `artifact-dir`, we say `target-dir` is the artifact directory and move everything else out into a "working directory". 
- The move can be gradual as these files are considered implementation details. 
- We avoid "yet another flag" problem of `--artifact-dir` because controlling of the location of the "working directory" is too low level for the CLI and should be reserved for config
- A big downside compared to `--artfact-dir` is that `target-dir` is not the final directory files are put under but the parent directory.  This is especially annoying for `cargo build` vs `cargo build --target`, dealing with profile names, etc

Potential names
- `working-dir` (arbitrarily chose this one to refer to the concept moving forward)
- `build-dir`

Cargo script would default its `target-dir` as its `working-dir`

This would need input from
- artifact-dir users
- cache / cache clean up users
- random other systems that dig into the target dir

**Comment by epage** [comment]
This would need an audit of ways we publicly treat the target dir as a working dir, like exposing [`CARGO_TARGET_TMPDIR`](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates)

**Comment by ensc** [comment]
Sharing sources over NFS adds another incompatibility when the ability to move `./target` out of the sources is eliminated.

Every modern buildsystem allows to keep sources and build results separated (and users and tools do not have problems with it).  I do not think that cargo should go the way back and enforce a fixed `./target` directory.

**Comment by kornelski** [comment]
I'm not suggesting to force it to always be `./target`. The `CARGO_TARGET_DIR` can continue to move this directory elsewhere.
The main point is to reduce severity of problems that the current high-churn high-volume content of this dir causes.

**Comment by epage** [comment]
We talked about this in today's Cargo team meeting.

Our care abouts include
- Moving intermediate artifacts away
  - Having a transition plan for tool authors
- Providing a profile-agnostic location for final artifacts
- Resolving the "elide target when host" within the final artifact path
- Having a coherent story for users, for example
  - Users today mostly interact with `target-dir` for dealing with final artifacts, so we likely want to preserve that
  - If we create a new directory for intermediate artifacts, we might not want to templatize `target-dir` as that will push people down one path moving intermediate artifacts and then we completely change it on the user what path they should go, rather than having a stable story for how to handle this

While we acknowledged the potential for user confusion with `CARGO_TARGET_TMPDIR`, we were fine with it not being associated with `target-dir`

The general shape of what we proposed in the meeting is...

### Shiny future

`target-work-dir`: Home of intermediate artifacts
- Goals:
  - Move intermediate artifacts out of source
  - Don't force a transition for what term/variable people reach out to to access final artifacts
- name is a placeholder
- No CLI flag for `target-work-dir`
- Precedence (note the lack of `target-dir`)
  - `target-work-dir` config
  - `target-work-dir` default (`"{cargo-cache}/target/{workspace-manifest-path-hash}"`)

`target-artifact-dir`: Home of final artifacts
- Goals:
  - Provide profile-agnostic path
  - Provide a way for users to opt-out of the target-platform from being elided when building for the host-platform
- name is a placeholder (e.g. do we want the `target` prefix)
- `--artifact-dir` CLI flag is added
- will error if `{platform}` or `{legacy-platform}` is not present with multi-target builds
- Precedence
  - `target-artifact-dir` config / cli
  - `target-dir` config / cli
  - `target-artifact-dir` default (`"{workspace-root}/target/{legacy-platform}/{profile}"`)

Legacy `target-dir`
- `--target-dir` is hidden on CLI.  Maybe shows up in man pages
- No templating supported

Other
- `CARGO_TARGET_TMPDIR` points within `target-work-dir`

### Path to Shiny Future

#### `target-work-dir`

In theory, we could trivially do this by
- locking `target-work-dir` and `target-dir` (if different)
- doing all work in `target-work-dir` and only when we do the "hardlink or cp" do we reference `target-dir`

Initial default is `"{workspace-root}/target"`

Template supports
- `{workspace-root}`
- `{cargo-cache}` (pointing to `CARGO_HOME` for now)
- `{workspace-manifest-path-hash}`

Steps
1. Implement
2. Call for testing
3. Stabilize with opt-in to new location
4. Call for testing
5. Switch to opt-out

Notes:
- Throughout those steps, we would work closely with tool authors to make sure they had a smooth transiti
- The opt-in/opt-out could either be
  - `.cargo/config.toml` `target-work-dir` field
  - A one-off environment variable if we don't want to stabilize the above yet

#### `target-artifact-dir`

Assumption: `target-work-dir` takes some pressure off of `target-artifact-dir`

Defaulted to final location (`"{workspace-root}/target/{legacy-platform}/{profile}"`)

Template supports
- `{workspace-root}`
- `{cargo-cache}` (pointing to `CARGO_HOME` for now)
- `{workspace-manifest-path-hash}`
- `{platform}`
- `{legacy-platform}` (elides host-target)

Needs all of the details in the tracking issue to be finalized.

### Contingencies

If `target-work-dir` takes more than N time (1 year?) to stabilize, then we re-evaluate approving rust-lang/rfcs#3371.  This is to try to balance the needs of the people who want something like rust-lang/rfcs#3371 now vs (1) the long-term inapplicability of that RFC and (2) the lack of stable "blessed" workflow for users (telling users to use solution X for several months and then telling them that is no longer "right" and they need to use solution Y).

### Alternatives

- Maintain `target-dir` as the "artifact base dir" and provide a query command (like `buck2 analyze`) to ask what the "artifact dir" is
- Don't expose `target-artifact-dir` in CLI, leaving it for more advanced uses
- Have only `target-dir` and `target-artifact-dir`
  - Downside: harder end-user transition as they reach out to `target-dir` for final artifacts.  While we eventually de-emphasize `target-dir`, those workflows will work just fine
- Have only `target-dir` and `target-work-dir`
  - Downside: access to final artifacts is still annoying
  - Only append final directories (`{platform}/{profile}`) if no templates are available
    - Downside: this would be confusing
    - We do this for `{dl}` but that is a more limited / off in the weeds use case rather than front and center for users
- Change the meaning of `target-dir` if `target-work-dir` or `target-artifact-dir` is present
  - This gets confusing

**Comment by epage** [comment]
Something we overlooked in the above analysis is other "artifacts".  In particular, I'm thinking of `cargo package` which places files in `$CARGO_TARGET_DIR/packages`.  I'm assuming at least the `.crate`s location is part our stable API.  We'd need to decide about the files laid out on disk next to it. 

Ways of solving this
- Like examples going to `examples/`, we could make the profile and target template dirs blank and always put this under `packages/`
- We could have a high level `{default}` variable that is artifact-specific.

**Comment by clarfonthey** [comment]
Just commenting here as I'm dealing with my own issues regarding the target dir, but personally, while it's nice for `target` to contain final build products only by default, I still will want all of these build products out of the target directory for the sake of excluding them from backups and snapshotting.

For some context, I use ZFS snapshots as a form of fast local backups; not long-term backups in case of hardware or extreme software failure, but decent short-term backups in case I accidentally delete a file or mess up an update. However, I explicitly go out of my way to exclude as many things as possible from auto-snapshotting that qualify as "cache" because they can very quickly clog up my disk if I'm not careful.

(Also: since snapshotting is a filesystem-level feature, I can't just say "don't save files of this type in snapshots" since snapshotting works by instantly copying the state of the FS, and doesn't copy files over like a long-term backup would.)

For example, today I just deleted 200 GiB of *snapshots* of target directories. Not the current target directories, but past versions of them from previous snapshots. Snapshots are good for incremental stuff like code because they're copy-on-write, but the contents of a binary are effectively random to any snapshotting tool and they'll end up being fully duplicated every time they're snapshotted, and that means you can end up with several times that amount of data in snapshots until everything eventually gets old enough to be deleted. The "effectively random" part also applies especially to the final products, since while crates that don't change won't change in their compiled artifacts, the final linked products definitely will.

So, as far as I'm concerned, moving the final build products back into the workspace without also having the option to keep them out effectively un-solves the problem that moving the target directory was meant to solve. After all, the final build products, modulo LTO (which isn't really going to happen for debug builds) will effectively be the same size as all the intermediate products, so, that means that about half the disk usage will not be saved. (I'm *extremely* approximating here; the point is that it's a considerable amount of the disk usage, even if it's not half. Even 10% of the size is still a lot when you consider that these are being multiplied across several snapshots.)

And note that yes, other languages like Node and Python *also* have this exact same problem, but I don't think that other languages' inability to solve this problem forgives Rust not solving it. Also, even though `node_modules` can be massive, hundreds of GiB of binary artifacts is pretty hard to beat.

I love the idea of keeping intermediate products deduplicated and in one place. I just don't want that to obscure the goal of having the final products *also* somewhere else too.

**Comment by kornelski** [comment]
> After all, the final build products, modulo LTO (which isn't really going to happen for debug builds) will effectively be the same size as all the intermediate products

No, the intermediate products are usually many many times larger. Not just double, they can be *1000&times; larger!* On the project I'm currently working, a debug build of 20MB executable creates 2300MB of junk in `target/`. 

There are often many duplicate copies of libstd and other dependencies in each `.rlib` file. There is a lot of duplication across code units. There's plenty of completely unused objects included in the dependencies, and stripped even without LTO (rust relies on `--as-needed` flag). There are also often separate copies for build dependencies, builds with `cfg(test)`, incremental build cache,

**Comment by epage** [comment]
@clarfonthey the plan calls for both `target-work-dir` and `target-artifact-dir` to be templated so you can move their content out.  It does not call out templating of `target-dir` as it calls for phasing that out.  If we wanted to templatize it as a convenience way of setting both of the above, we'd likely want to wait for the above so we set the precedence for what people are generally expected to work with, rather than shifting expectations around on the user.

**Comment by mathstuf** [comment]
> I still will want all of these build products out of the target directory for the sake of excluding them from backups and snapshotting.

FWIW, I've had decent results with `target` being a symlink to somewhere that is not subject to backups/snapshotting. `cargo clean` will remove the symlink, but everything else I've used is largely fine. Note that this puts intermediate and final artifacts into the same bucket.

**Comment by nazar-pc** [comment]
I'm in the same boat as @clarfonthey, but with BTRFS snapshots, which I create every 15 minutes and then stream to longer-term storage. My debug builds at easily 2.5G and I clean up `target` that is 300-700G basically every week.

I created `~/.cache/cargo/{git,registry,target}` for this reason, where `~/.cache` is in a separate subvolume that is not subject for snapshotting/backups. `~/.cargo/{git,registry}` are symlinks now (still hoping [cargo starts respecting XDG one day](https://github.com/rust-lang/cargo/issues/14125)) because they also grow to substantial sizes (currently 1.08M files and 24.1G together).

Proposed separation (especially templating for both new options) should work nicely for such use case, `CARGO_TARGET_DIR` in `.profile` has major consequences for build times when jumping between projects.

Excited!

**Comment by Nemo157** [comment]
Personally I also don't care about having the artifacts copied anywhere, I access them via `cargo run` etc. directly from the work-dir anyway. It'd be nice if we could do something like `target-artifact-dir = false` to just disable it.

**Comment by ranger-ross** [comment]
> Personally I also don't care about having the artifacts copied anywhere, I access them via cargo run etc. directly from the work-dir anyway. It'd be nice if we could do something like target-artifact-dir = false to just disable it.

My understanding is the the `target-artifact-dir` will default to `{workspace-root}/target/{legacy-platform}/{profile}` so it will still be contained in the `target` directory by default. If you are only running your project via `cargo run` I don't think you need to worry about the exact location of the binary in the `target` dir?

One question I have is about the handling of building for the host platform. 
Running `cargo build` & `cargo build --release` will build artifacts to `target/debug/` and `target/release/` dirs respectively. 
Changing this would be a fairly big user facing change and will result in a lot of churn in the ecosystem.
If I am reading [this](https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917) correctly, I am assuming that this default behavior will stay the same unless the user explicitly opts to use `target-artifact-dir` to change it. 
The new `target-artifact-dir` would however change if a `--target` is passed to the `cargo build`.

Is this correct?

**Comment by Nemo157** [comment]
> If you are only running your project via `cargo run` I don't think you need to worry about the exact location of the binary in the `target` dir?

One thing I do care about the artifact locations is that they are _not_ in a sub-dir of the repository. Having large temporary artifacts mixed in with the source code complicates backups and filesystem snapshots. This is why I would prefer to disable the target-artifact-dir and just leave all intermediate and final artifacts somewhere in `~/.cache/cargo`.

**Comment by ranger-ross** [comment]
Ahh I see. 

I think this should be doable with the current proposal by setting `target-artifact-dir` to `~/.cache/cargo/artifacts`, right?

**Comment by ranger-ross** [comment]
I had a brief [conversion](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/Splitting.20build.20and.20artifact.20directories) with @epage  in Zulip and I would like to help implement this.

My plan is to follow the path laid out in this [comment](https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917).

@rustbot claim

---

> `target-build-dir`: Home of intermediate artifacts
> [...]
> name is a placeholder

Regarding the naming of `target-build-dir`, I believe this name is fine. 
I was thinking about possible alternatives:
* `target-intermediate-dir`
* `target-work-dir`
* `target-interim-dir`

I am not too opinionated on the naming, but I figured I would bring it up sooner rather than later.

**Comment by epage** [comment]
I gave those names largely as placeholders.  I have mixed opinions on whether to have the `target-` prefixes.  In the medium term, they connect it to `target-dir`.  In the long term, we want to phase out `target-dir` and `target-artifact-dir` doesn't line up well with `--artifact-dir`.

Overall though, we can pick names and go with it while its unstable and then address it while we work to stabilize it.

**Comment by ranger-ross** [comment]
A question that arose about the behavior of path templating in #15236.

The path variables `{workspace-root}` and `{cargo-cache}` refer to specific directories while `{workspace-manifest-path-hash}` is simply a part a path.

It does not make logical sense to allow `{workspace-root}` and `{cargo-cache}` to not be the first variable in the path template.
ie. 
```toml
# Assuming the project is at `/home/myuser/project`
build-dir = "{workspace-root}/foo" # All good! Expands to `/home/myuser/project/foo`
build-dir = "/foo/{workspace-root}/bar" # Not valid. Would expand to `/foo/home/myuser/project/bar` ?
```

This is different than `{workspace-manifest-path-hash}` that can be place in the middle of the path template
```toml
build-dir = "/foo{workspace-manifest-path-hash}/bar" # Expands to `/foo/34/f9d02eb8411c05/bar`
```

The question is what should Cargo do when it receives an invalid path template in `build.build-dir`?

Some potential options are 
* Fail early and display an error to user.
* Ignore the everything before the `{workspace-root}`
* Do nothing and allow it to fail later

---

This is just one of quiet a few invalid scenarios that are possible with path templating.
A few others that I can think of are:
```toml
# Using the same variable twice
build-dir = "{workspace-root}/{workspace-root}" 

# Using workspace-root and cargo-cache together
build-dir = "{workspace-root}/{cargo-cache}" 
```

**Comment by epage** [comment]
@ranger-ross we discussed a few of the open questions in today's cargo meeting.
- We'd like to rename `workspace-manifest-path-hash`.  `manifest` at least is extraneous.
  - Dropping down to `hash` may confuse users who don't realize moving their workspace gets them a new build-dir
  - `workspace-hash` may have some of the same problem, just not as bad
  - `path-hash` leaves it slightly ambiguous of whether its workspace, package, or current-dir path
  - `workspace-path-hash` might be a decent middle ground
  - **Any thoughts on which direction to go?**
- We'd like the hash to include the Cargo version, minus the release channel.
  - This ensures no one can accidentally rely on our hash implementation
  - If/when we get build-dir garbage collection, we'd be able to clean up build-dirs for old versions of Cargo very easily without having to do GC within each build-dir
  - We need to exclude the release channel so that every nightly doesn't create a whole new target directory, exploding discus space (generally the incremental cache)

We still need to discuss symlinks and error handling.

**Comment by ranger-ross** [comment]
I think I lean towards `workspace-path-hash` as well.
I agree that `hash` and `path-hash` are too vague.

> * We'd like the hash to include the Cargo version, minus the release channel.     
>   * This ensures no one can accidentally rely on our hash implementation
>   * If/when we get build-dir garbage collection, we'd be able to clean up build-dirs for old versions of Cargo very easily without having to do GC within each build-dir
>   * We need to exclude the release channel so that every nightly doesn't create a whole new target directory, exploding discus space (generally the incremental cache)
> 

Makes sense to me 👍

**Comment by epage** [comment]
@ranger-ross we discussed more of the open questions
- We are deferring #15340 because we are concerned about tools like [`cargo-sweep`](https://crates.io/crates/cargo-sweep) that would need to know all of the `build-dir`s (sorry)
- We do want `build-dir` included in `cargo metadata`  but only the current `build-dir`.  We should make it clear that the content is unspecified.
- We decided to go ahead and go for resolving symlinks.  There was not a strong feeling on this.

**Comment by epage** [comment]
@ranger-ross thanks for your work (and @weihanglo's reviews while I've been absent).

In updating the task list for what was done, I realized I didn't have erroring for unmatched `{` / `}` on the task list.  I also suggested we append a `closest_msg` to the unknown variable error / list all variables to help people in navigating this while testing it.

**Comment by epage** [comment]
Waiting on #15418 to hit a nightly before doing a Call for Testing.

Places we should specifically reach out to about this include:
- #9661 as that has people interested in the target dir location and may want a heads up that changes to it are coming down the pipe)
- #11156 as that will be resolved with this being stabilized
- [cargo-sweep](https://crates.io/crates/cargo-sweep) as they will be impacted by users switching `build-dir`
- [cargo-sweep](https://crates.io/crates/cargo-cache) as they will be impacted by users switching `build-dir`
- [snapbox](https://crates.io/crates/snapbox) as it finds tries to find a final artifact based on the location of an intermediate artifact
  - **This impacts cargo's testsuite**
- [assert_cmd](https://crates.io/crates/assert_cmd) as it finds tries to find a final artifact based on the location of an intermediate artifact
- [#t-cargo/build-integration](https://rust-lang.zulipchat.com/#narrow/channel/334885-t-cargo.2Fbuild-integration) as that might have people doing ... strange things (or could benefit)

Things we should remember to include:
- Does the division of intermediate / final artifact work?
- Do a test run with `cargo clean` being done first and do a full validation
  - `cargo clean` is important for test suites to not access stale binaries when they inferred final artifacts from intermediate artifacts
- Ask to test variable substitution
- Warn that the hash is unspecified and we are considering making it change with every cargo release which allows auto-cleanup of old releases when combined with #13136
- Give a heads up that `cargo-cache-home` maps to `CARGO_HOME`, its just named for when we add support for XDG and its location can be changed between releases.

**Comment by epage** [comment]
### Call for testing

`build-dir` is now available in nightly and we're looking for feedback to see if this can be stabilized!

This allows users to move intermediate build artifacts into a location separate from final artifacts.  Use cases include
- Moving build artifacts to a location with faster IO
- Making it easier to periodically find and delete all of the large build caches (see also #13136)
- Avoid being picked up by filesystem snapshotting

This can be set in user config (e.g. `~/.cargo/config.toml`) without workspaces conflicting with each other due to the support for templating

e.g. `~/.cargo/config.toml`
```toml
[build]
build-dir = "{cargo-cache-home}/build/{workspace-path-hash}"
```

Future possibilities
- Change the default `build-dir` location (e.g. to the value above)
- Change the layout of files within `build-dir` (#15010)

#### Basic steps:

Requirements:
- 1.88.0-nightly (092a284ba 2025-04-13) or newer

See the [unstable docs](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-dir) for what is supported

Please test your workflows with `-Zbuild-dir`, including:
- With *and* without setting `build.build-dir`
- Explicit and default `build.target-dir`
- Various commands that write different kinds of data/locations within `target-dir` (`cargo check`, `cargo build`, `cargo package`, `cargo doc`, etc)
- Tests (including after a `cargo clean`)
- Tools and processes that may touch `target-dir` (e.g. CI caching, `cargo-cache`, `cargo-sweep`)

When testing, please consider
- Does the division of files between `build-dir` and `target-dir` make sense?

**Notes:**
-  `{manifest-path-hash}`s hash is not guaranteed to be stable between releases and, in fact, we may go as far as also hashing the Cargo version as that blends well with #13136.
- `{cargo-cache-home` evaluates to `CARGO_HOME` for now but is named as such to be prepared for #1734

**Please leave testing reports on this issue; feel free to open dedicated issues for bugs**

**Comment by asomers** [comment]
Test report
=========

I tested this feature with the mdconfig, libnv-rs, and fsx crates on FreeBSD 15 amd64, as well as with a complicated unpublished crate.  It's a big step forward!

**What worked:**
- Using either `.cargo/config.toml` or `$HOME/.cargo/config.toml`
- Using either the `CARGO_BUILD_BUILD_DIR` variable or an entry in `.cargo/config.toml`
- Using the `{cargo-cache-home}`, `{workspace-path-hash}`, or `{workspace-root}` templating variables for `build-dir`
- Omitting `-Zbuild-std` restores legacy behavior
- Executing tests with a non-default `build-dir`.
- `cargo doc` could build the tests, though it didn't put the html files in `build-dir`.
- Running `cargo +nightly -Zbuild-dir test` in a workspace with a `rust-toolchain.toml` file, as long as I explicitly specified `+nightly`.
- Building crates with build scripts.

**What did not work**
- The `assert_cmd` crate did not work, as noted by @epage .
- Using templating variables for `target-dir`.  For my workflow, that's important.  I share the same source directory (over NFS) between VMs running different operating systems, so I really do want to move finished binaries to alternate locations.  Moving html files generated by `cargo-doc` would be good too.


**Another feature request**
While `{workspace-path-hash}` works well for automated tools, it's not so great for human use.  It would be convenient for me, though not strictly necessary, to have a template variable that would expand to the base name of the workspace root.  That would make it easier for me to clean up my temporary files when my HDD fills up and I have two dozen crates with builds lying around.

**Comment by epage** [comment]
Thanks for giving it a spin!

> cargo doc could build the tests, though it didn't put the html files in build-dir.

the html files were considered a "final artifact" so we kept them in `target-dir`. This is a problem of use cases
- `cargo doc --no-deps` is important for publishing documentation (final artifact)
- `cargo doc` is important for development

> Using templating variables for target-dir. For my workflow, that's important. I share the same source directory (over NFS) between VMs running different operating systems, so I really do want to move finished binaries to alternate locations. 

We plan to add an `artifact-dir` which would be templated and phase out `target-dir`. Technically templating would be a breaking change for `target-dir` though it is unlikely to affect anyone. The question would be if templating for `target-dir` until we have `artifact-dir` would be worth the effort, including the compatibility analysis.

> While {workspace-path-hash} works well for automated tools, it's not so great for human use. It would be convenient for me, though not strictly necessary, to have a template variable that would expand to the base name of the workspace root. That would make it easier for me to clean up my temporary files when my HDD fills up and I have two dozen crates with builds lying around.

While you can look it up with `cargo metadata`, I agree that prefixing the hash with the basename (or offering a variable for the base name) would make this a lot more convenient for humans.

**Comment by asomers** [comment]
> Thanks for giving it a spin!

> We plan to add an `artifact-dir` which would be templated and phase out `target-dir`. Technically templating would be a breaking change for `target-dir` though it is unlikely to affect anyone. The question would be if templating for `target-dir` until we have `artifact-dir` would be worth the effort, including the compatibility analysis.

Ahh, that sounds like it would work for me.

> 
> > While {workspace-path-hash} works well for automated tools, it's not so great for human use. It would be convenient for me, though not strictly necessary, to have a template variable that would expand to the base name of the workspace root. That would make it easier for me to clean up my temporary files when my HDD fills up and I have two dozen crates with builds lying around.
> 
> While you can look it up with `cargo metadata`, I agree that prefixing the hash with the basename (or offering a variable for the base name) would make this a lot more convenient for humans.

Is it possible to use `cargo metadata` within .cargo/config.toml?  Alternatively, prefixing the hash with the basename would also work for me.

**Comment by xd009642** [comment]
# Test report

I tested this feature by using tarpaulin with build dir on a few projects. Using the llvm and ptrace engine (as both have slightly different requirements in terms of artefacts). It all worked as expected, likely due to tarpaulin mostly relying on cargo_metadata to get paths to artefacts instead of relying on any structure of the target directory. 

* Using either .cargo/config.toml
* Using either the CARGO_BUILD_BUILD_DIR variable or an entry in .cargo/config.toml
* Using the {cargo-cache-home}, {workspace-path-hash}, or {workspace-root} templating variables for build-dir
* Omitting -Zbuild-std restores legacy behavior
* Executing test and doc-test coverage with a non-default build-dir.
* I also tried slightly messing up the template vars or making up ones and error messages were as expected (suggesting ones with close string distance, unexpected var with quotes around the name  etc.

How I messed up (what didn't work)

* Initially I was lazy and didn't use `$HOME` but `~`, which I should know wouldn't work (but it's early).  This resulted in a `~` directory in my project which is understandable. Maybe something that warrants a compiler warning 🤷

**Comment by riverar** [comment]
I'm having a hard time understanding the new feature. The docs, your latest comment, etc. use the term intermediate but I can't find what the final state is. Are these intermediate files copied somewhere after the crate is compiled? After the entire build is complete? Other?

Your last comment ("This just leaves the use case...") seems to imply you covered the "I want to deploy files to the target dir" scenario but I don't understand how to use this feature to do so.

**Comment by kornelski** [comment]
1. The requirement to add `-Zbuild-dir` makes it harder to test the feature, since it's hard to control all the places that invoke `cargo`, such as editors that run `cargo check` or nested build systems and `build.rs` scripts that invoke Cargo themselves. For experimental `rustc` features I can set `rustflags` in `~/.cargo/config.toml`, but Cargo doesn't have such thing for itself AFAIK.

2. I was surprised that `{workspace-path-hash}` is not a single path component, but has a 2-char directory prefix. I assumed it'd be alphanumeric single component, since it's a *hash* of a path, not a path itself. Hopefully not a big deal, since I won't need to navigate these directories manually, just unexpected.

3. Can this be used as an opportunity to always add the target triple to the path? I've ended up with both `~/Library/Caches/Cargo/{workspace-path-hash}/debug` and `~/Library/Caches/Cargo/{workspace-path-hash}/my-host-triple/debug` dirs. Since these directories are out of users' view, they could be made predictable instead of short for user convenience. The irregularities with the optional triples and `--profile=dev` being `debug` in the path were annoying when building `cargo-xcode` and `cargo-deb`.

**Comment by epage** [comment]
> The requirement to add -Zbuild-dir makes it harder to test the feature, since it's hard to control all the places that invoke cargo, such as editors that run cargo check or nested build systems and build.rs scripts that invoke Cargo themselves

‘-Z` flags can be set through the config system, see https://doc.rust-lang.org/cargo/reference/unstable.html

> Can this be used as an opportunity to always add the target triple to the path?

Maybe eventually
- We should keep this specific change focused and iterate in steps past it
- It will help if we first make more explicit our current lack of guarantees around the build dir before making further changes. In fact, changing the layout is the reason ranger-ross has been contributing these changes. Their focus is on changing the layout in prep for a shared cache.

**Comment by zopsicle** [comment]
# Test report

This feature appears to work great for my use case detailed here: https://github.com/rust-lang/cargo/issues/6100#issuecomment-2817178766! Using the following commands:

```bash
cargo build \
    -Zbuild-dir \
    --config 'build.build-dir = ".build/cargo"' \
    --config 'build.target-dir = ".build/ush"' \
    --package ush

cargo build \
    -Zbuild-dir \
    --config 'build.build-dir = ".build/cargo"' \
    --config 'build.target-dir = ".build/ush-lsp-server"' \
    --package ush-lsp-server
```

where ush-lsp-server depends on ush with additional features; feature unification is correctly inhibited by building separately; overwriting of artifacts is correctly inhibited by specifying different `target-dir`s; and repeated compilation is correctly avoided by specifying the same `build-dir`.

`cargo test` works properly without `target-dir`. It seems to write no artifacts, as expected.

`cargo doc` works properly: the rustdoc output is written to `target-dir`.

**Comment by Nemo157** [comment]
Having `{cargo-cache-home}` resolve to `CARGO_HOME` feels like it exposes the possibility of conflicts between cargo's future use of that directory and arbitrary user config, e.g. as an absurd setting `build-dir = "{cargo-cache-home}/registry/src/index.crates.io-1949cf8c6b5b557f"`.

I'm not sure if there's really anything nice that can be done about this, just thought it'd be worthwhile to note as I don't recall it being mentioned before. (I've long wondered what would happen if cargo integrated native audit support and wanted to use `CARGO_HOME/audit.toml` when `cargo-audit` has already claimed that as its own).

**Comment by teohhanhui** [comment]
> I'm not sure if there's really anything nice that can be done about this

Don't expose `cargo-cache-home` but a `cargo-build-home` maybe?

**Comment by epage** [comment]
> Having {cargo-cache-home} resolve to CARGO_HOME feels like it exposes the possibility of conflicts between cargo's future use of that directory and arbitrary user config, e.g. as an absurd setting build-dir = \"{cargo-cache-home}/registry/src/index.crates.io-1949cf8c6b5b557f\".

I'm not seeing how this is a new problem. Someone can already construct a path like this without variables. Granted, this makes it easier but I could see offering a `{home}` which is the minimum for making this easier. We also don't prevent top-level `../` so if we had an alternate location, users could still reach it.

As for future use, I'm not quite sure what you mean but when we make `CARGO_CACHE_HOME`, `registry/` will be a part of it, so this possibility won't change. `audit.toml` wouldn't be though.

**Comment by Nemo157** [comment]
> I'm not seeing how this is a new problem. Someone can already construct a path like this without variables.

I think it's more obvious that using a path like `build-dir = "~/.cargo/builds"` is reaching into internal implementation details and might cause issues (though, apparently not for a lot of third-party subcommand authors). Using a variable that seems to be explicitly for this usecase makes it less likely for users to consider it.

> As for future use [...]

I mean that a setting for `build-dir` that is ok now may conflict with new subdirectories/files introduced by cargo in future versions. This is not entirely theoretical, my previous setting of `target-dir = "~/.cargo/target"` ran into this conflict with `-Zscript`.

---

The only thought I have to actually doing something about this is either documentation about what cargo expects from subdirectories of `CARGO_HOME`/`CARGO_CACHE_HOME, or having a variable more like `$CARGO_HOME/user/` which is explicitly set aside for arbitrary files outside cargo's control.

**Comment by ensc** [comment]
miri (13e8790949 2025-05-04) fails to run with

```
$ cargo miri clean
$ cargo miri test
Preparing a sysroot for Miri (target: x86_64-unknown-linux-gnu)... fatal error: failed to build sysroot: failed to read cargo out dir

Caused by:
    No such file or directory (os error 2)
```

when

```
[build]
build-dir = "{cargo-cache-home}/build/{workspace-path-hash}"

[unstable]
build-dir=true
```

is set in `~/.cargo/config.toml`.

After disabling these lines, miri can create the sysroot and works when these lines are reenabled.

**Comment by epage** [comment]
I've opened an issue at RalfJung/rustc-build-sysroot#25

**Comment by RalfJung** [comment]
Various crates use the CARGO_TARGET_DIR env var set by cargo during a `cargo test` build to store some files, e.g. ui_test. Those are typically not artifacts, they are intermediate build files and just serve as cache. With the intention behind CARGO_TARGET_DIR changing, should those crates use a different location instead?

**Comment by Skgland** [comment]
> Various crates use the CARGO_TARGET_DIR env var set by cargo during a `cargo test` build to store some files, e.g. ui_test. Those are typically not artifacts, they are intermediate build files and just serve as cache. With the intention behind CARGO_TARGET_DIR changing, should those crates use a different location instead?


There is `CARGO_TARGET_TMPDIR` for integration tests and benchmarks, would that work?

> CARGO_TARGET_TMPDIR — Only set when building integration test or benchmark code. This is a path to a directory inside the target directory where integration tests or benchmarks are free to put any data needed by the tests/benches. Cargo initially creates this directory but doesn’t manage its content in any way, this is the responsibility of the test code.
> 
> - https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates

**Comment by RalfJung** [comment]
Ah, yes that sounds like the right env var, thanks. Since when is that available?

**Comment by ranger-ross** [comment]
>  Since when is that available (to avoid bumping MSRV too much)?

Seems to have been merged in #9375 and stablized in [1.54.0](https://doc.rust-lang.org/nightly/cargo/CHANGELOG.html#cargo-154-2021-07-29)

**Comment by U007D** [comment]
This CfT has been added to TWiR Issue 602.

You may now remove the call-for-testing label. Please feel free to re-add the label if you wish this CfT to appear again in a future issue.

**Comment by epage** [comment]
I propose that we stabilize [`build-dir` config and its template variables](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-dir).  artifact-dir is being tracked in #6790.  Phasing out `target-dir` is blocked on both of these efforts.  Changing the default for `build-dir` from `target-dir` to something else is left for a future exercise.

`build-dir` controls where intermediate build artifacts live.  By stabilizing this we are more explicitly than [our docs](https://doc.rust-lang.org/cargo/reference/build-cache.html) locking down what is an intermediate or final build artifact.  The tricky part to this that this can also be about intent, e.g. building docs to verify them and not to look at them.

For more details, see
- [docs](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-dir)
- [testsuite/build_dir.rs](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/build_dir.rs)


This allows users to move intermediate build artifacts into a location separate from final artifacts.  Use cases include
- Moving build artifacts to a location with faster IO
- Making it easier to periodically find and delete all of the large build caches (see also #13136)
- Avoid being picked up by filesystem snapshotting
- Be clearer about our compatibility guarantees, especially in a lead up to #15010

Of particular note:
- A concern was raised that we exposed a variable for `{cargo-cache-home}` (https://github.com/rust-lang/cargo/issues/14125#issuecomment-2826700293) because we could put new content there and stomp on users content.  I lean towards accepting that risk
  - We could have an opinionated fully specified path that includes the workspace-path-hash, etc but I think users will want some more flexibility than that and its more opaque which can be de-empowering for users / not acting as documentation / a starting off point for users to provide their own
  - We could have a `{cargo-build-home}` which maps to `CARGO_CACHE_HOME/build` that is reserved for users.  However, if/when we change the default, we won't be able to express it in terms of `cargo-build-home` and would either need to add `cargo-cache-home` or have an opinionated, fully specified path, see above.
- Test libraries have done hacks to lookup `bin`s at runtime, rather than using `CARGO_BIN_EXE_<name>` (e.g. `cargo-test-support` before #15692).  Like with #15692, I suspect in a lot of cases we can transition away from these.  For anything remaining, it would be good to get that feedback from users before making the variables available at runtime (see also [#t-cargo > cargo_bin_exe and tests @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/cargo_bin_exe.20and.20tests/near/513638220)) 
- We did a call for testing but I'm most worried about users posting in #9661 for which we didn't get much of a response.  As this is still opt-in for now, we can work with the ecosystem through issues over time
- For users seeing the name, `build` can refer to builds, `cargo build`, or `package.build` and depending on what contexts get created, this could end up being confusing (see next item).
  - However, this still seems like the best name of the bunch we had considered.
- `CARGO_TARGET_TMPDIR` will point to within `build-dir`.
  - We could transition users to an alternative name but I figure this could be deferred for now.  We'd have to figure out naming, e.g. `CARGO_BUILD_TMPDIR` seems a natural parallel but (1) that could be confused for a scratchpad directory for build scripts that was requested at one point and (2) this lives in `build-dir` but is for use in tests.
- The `build-dir` template errors out on unknown variables or unmatched pairs
- When determining the workspace-path-hash, we resolve symlinks though the details of the hash are not guaranteed and can be changed

We've tried to be proactive in our communication of this feature, including
- holmgr/cargo-sweep#140
- matthiaskrgr/cargo-cache#143
- taiki-e/cargo-llvm-cov#432
- xd009642/tarpaulin#1721
- rust-lang/rust-analyzer#20150

On stabilization, we'll need to
- Create an issue for changing the default `build-dir`

@rfcbot fcp merge

**Comment by ensc** [comment]
> On stabilization, we'll need to
> 
>     * Create an issue for changing the default `build-dir`

and allow substitution of variables for `target-dir`

**Comment by epage** [comment]
@ensc in our plan (https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917) we called out no templating for `--target-dir` but instead we'd phase it out in favor of the build-dir and artifacts-dir which would support templating.

**Comment by ehuss** [comment]
What is the long-term plan for how `cargo clean` is intended to work? It seems a little surprising to me that it deletes the entire global cache (if not segregated via workspace). How will the user have control over local artifacts versus the global build directory?

It seems that this doesn't provide any concessions to tools that need to access these files, and seems to break their behavior. For example, this is incompatible with building rustc. What is the expectation for how tools are supposed to access these files? What is the scope of tools breakage?

I'm not quite understanding the relationship with https://github.com/rust-lang/cargo/issues/5931 (per-user global cache). Is the intent that this takes over that job? That is, setting `build.build-dir` to a global location resolves that? Or is the intent for addressing #5931 be a *replacement* for some uses of build-dir?

**Comment by ehuss** [comment]
Does `workspace-path-hash` have any handling for hash collisions? It looks like a very small two-character hash.

**Comment by epage** [comment]
> What is the long-term plan for how cargo clean is intended to work? It seems a little surprising to me that it deletes the entire global cache (if not segregated via workspace). How will the user have control over local artifacts versus the global build directory?

Note that in the FCP proposal, sharing of `build-dir`s or replacing #5931 were not given as motivations.  We are not aiming for that use case with this and solving problems like locking or `cargo clean` with a shared `build-dir` are out of scope of this and should be left to #5931.  That said, sharing a `build-dir` does improve on the state of sharing a `target-dir` in that only intermediate artifacts, which mostly have a hash in them, are shared, and not final artifacts.

This is indirectly related to #5931 in that this is a step towards #15010, by giving a way for people to tell what will be broken by that ahead of time, which is a step towards #5931, as that will make it easier to cache artifacts, among other motivations.

> It seems that this doesn't provide any concessions to tools that need to access these files, and seems to break their behavior. For example, this is incompatible with building rustc. What is the expectation for how tools are supposed to access these files? What is the scope of tools breakage?

These tools are accessing implementation details of Cargo and the fact that using this feature breaks those tools is a positive outcome of this as it helps raise awareness of this dependence on implementation details for when we move forward with #15010 (and when we change the default `build-dir`).

From https://doc.rust-lang.org/cargo/reference/build-cache.html

> Cargo also creates several other directories and files needed for the build process. Their layout is considered internal to Cargo, and is subject to change. Some of these directories are:

For tools that wish to continue to access implementation details at their own risk, we do expose `build-dir` in `cargo metadata`
https://github.com/rust-lang/cargo/blob/eabb4cd923deb73e714f7ad3f5234d68ca284dbe/tests/testsuite/metadata.rs#L4314-L4328

> Does workspace-path-hash have any handling for hash collisions? It looks like a very small two-character hash.

`workspace-path-hash` evaluates to `{hash[0..2]}/{hash[2..]}`
https://github.com/rust-lang/cargo/blob/eabb4cd923deb73e714f7ad3f5234d68ca284dbe/src/cargo/util/context/mod.rs#L675-L679

This was taken from the now-closed RFC 3371, see https://github.com/poliorcetics/rfcs/blob/d549cb12889d5a2c4a4dd64b5a8a0e3ee7b49aa0/text/3371-cargo-target-dir-templates.md#handle-possibly-thousands-of-directories-in-a-single-templated-cargo_target_dir-path

The layout of `workspace-path-hash` is an implementation detail and we can change it as needed.

**Comment by epage** [comment]
> It seems that this doesn't provide any concessions to tools that need to access these files, and seems to break their behavior. 

I forgot to add that we've been proactively reaching out to the libraries and tools that we are aware of that would be impacted by this change.

**Comment by weihanglo** [comment]
Given the interaction with `cargo clean` is called, should we document the behavior of deleting the entire global caches, and this feature is not a solution of #5931?

**Comment by ranger-ross** [comment]
> Given the interaction with `cargo clean` is called, should we document the behavior of deleting the entire global caches, and this feature is not a solution of #5931?

I think that would be reasonable. target-dir has the same problem but I think it's a bit easier to misuse due to `{cargo-cache-home}` being opaque.

One other idea I had was to potentially have a cargo lint if build-dir (and potentially target-dir) are set to know problematic patterns like `{cargo-cache-home}`, `~`, and `/`.

**Comment by epage** [comment]
For me the situation is no different than target-dir. If we document shared target-dir, we should update it. If not, that is a separate concern from this effort.

**Comment by weihanglo** [comment]
That is fair. I'll check my box.

@rfcbot reviewed

**Comment by kornelski** [comment]
I'm afraid that all the tooling which uses `cargo metadata`, Cargo's internal API, or invokes `cargo build` command by itself is undertested, because setting `CARGO_BUILD_BUILD_DIR` or passing `-Zbuild-dir` is simply ignored in these scenarios.

`CARGO_BUILD_BUILD_DIR=x cargo asm -Zbuild-dir` doesn't use the `build-dir` feature.
`CARGO_BUILD_BUILD_DIR=x cargo deb -- -Zbuild-dir` doesn't use the `build-dir` feature.
`CARGO_BUILD_BUILD_DIR=x cargo xcode` doesn't use the `build-dir` feature, and doesn't support passing `-Zbuild-dir`
`CARGO_BUILD_BUILD_DIR=x cargo tauri dev` doesn't use the `build-dir` feature, and doesn't support passing `-Zbuild-dir`

**Comment by ranger-ross** [comment]
While preparing the stabilization PR I noticed some of the packaging tests started failing ([this one](https://github.com/rust-lang/cargo/blob/9b296973b425ffb159e12cf3cd56580fd5c85382/tests/testsuite/package.rs#L5683) for example)

I noticed we had one less `[COMPILING] whatever 0.0.1 ...` for a few of the `cargo package` related tests.
I did some digging and I came across a behavior change that was not discussed or noticed in the original implementation pull request.

---

During the post package "verify" step the `.crate` is untarred and built to verify it is able to compile.
Prior to splitting `target` and `build-dir`, during the build if `target-dir` was not set the build target directory would be set as `target/package/{name}-{version}/target`, due to [`Workspace::root()`](https://github.com/rust-lang/cargo/blob/9b296973b425ffb159e12cf3cd56580fd5c85382/src/cargo/core/workspace.rs#L422) calling `PathBuf::parent()` on the untarred manifest which is inside of the `target/package/{name}-{version}` directory.

This causes all of the dependencies to be re-built as the `target-dir` (which is now `target/package/{name}-{version}/target`) is empty.
It also means that for every crate in the workspace, will recompile all dependencies during package verification.

It's notable that by setting `CARGO_TARGET_DIR` this behavior changes. 
The workspace `target-dir` will be used during package verification meaning that if a crate has already been compiled with an unchanged fingerprint Cargo will skip compiling.

Below is a simplified example.

```sh
cargo new package-rebuild-test
cd package-rebuild-test

# Add a depedency (using tokio with full features to make it obvious when rebuilds happen)
cargo add tokio --features full

# This will always rebuild all dependencies
cargo package --allow-dirty

# This only builds dependencies on the first run
CARGO_TARGET_DIR=foo cargo package --allow-dirty
```

With the build-dir changes, the default behavior changes to as `CARGO_TARGET_DIR` was always set, expect we will be using the `build-dir` instead of the `target-dir`.  

---

I would argue this is a positive change as it reduces the amount of recompiling needed during packaging/publishing.

**Comment by epage** [comment]
> I would argue this is a positive change as it reduces the amount of recompiling needed during packaging/publishing.

If so, we should have an explicit PR to set this as the behavior, including tests.

I always took the current behavior as intended to ensure the correctness in place of build caches being poisoned in some way.  At least for local packages, they would still be fingerprinted differently than the original workspace members, do maybe less so.  I wonder if there are also problems with conflicting final artifacts.

With us stabilizing workspace publishing, I would be open to us sharing a build directory between published packages.  A part of me worries about sharing final artifacts.
